File: src/org/elixir_lang/structure_view/node_provider/Used.java
package org.elixir_lang.structure_view.node_provider;

import com.intellij.icons.AllIcons;
import com.intellij.ide.util.ActionShortcutProvider;
import com.intellij.ide.util.FileStructureNodeProvider;
import com.intellij.ide.util.treeView.smartTree.ActionPresentation;
import com.intellij.ide.util.treeView.smartTree.ActionPresentationData;
import com.intellij.ide.util.treeView.smartTree.TreeElement;
import com.intellij.openapi.actionSystem.Shortcut;
import com.intellij.openapi.util.Pair;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.PsiReference;
import com.intellij.util.IncorrectOperationException;
import org.apache.commons.lang.math.IntRange;
import org.elixir_lang.psi.ElixirAccessExpression;
import org.elixir_lang.psi.QualifiableAlias;
import org.elixir_lang.psi.call.Call;
import org.elixir_lang.structure_view.element.*;
import org.elixir_lang.structure_view.element.modular.Module;
import org.jetbrains.annotations.NonNls;
import org.jetbrains.annotations.NotNull;

import javax.swing.*;
import java.util.*;

import static com.intellij.openapi.util.Pair.pair;
import static org.elixir_lang.psi.impl.ElixirPsiImplUtil.stripAccessExpression;
import static org.elixir_lang.psi.impl.call.CallImplKt.finalArguments;
import static org.elixir_lang.psi.impl.call.CallImplKt.macroChildCalls;
import static org.elixir_lang.structure_view.element.modular.Module.addClausesToCallDefinition;

public class Used implements FileStructureNodeProvider<TreeElement>, ActionShortcutProvider {
    /*
     * CONSTANTS
     */

    @NonNls
    public static final String ID = "SHOW_USED";
    private static final String USING = "__using__";

    /*
     * Static Methods
     */

    private static Collection<TreeElement> filterOverridden(@NotNull Collection<TreeElement> nodesFromChildren,
                                                            @NotNull Collection<TreeElement> children) {
        Map<Pair<String, Integer>, CallDefinition> childFunctionByNameArity = functionByNameArity(children);
        Collection<TreeElement> filtered = new ArrayList<>(nodesFromChildren.size());

        for (TreeElement nodeFromChildren : nodesFromChildren) {
            if (nodeFromChildren instanceof CallDefinition) {
                CallDefinition callDefinition = (CallDefinition) nodeFromChildren;

                // only functions work with defoverridable
                if (callDefinition.time() == Timed.Time.RUN) {
                    Pair<String, Integer> nameArity = pair(callDefinition.name(), callDefinition.arity());

                    if (childFunctionByNameArity.containsKey(nameArity)) {
                        continue;
                    }
                }
            }

            filtered.add(nodeFromChildren);
        }

        return filtered;
    }

    public static Map<Pair<String, Integer>, CallDefinition> functionByNameArity(@NotNull Collection<TreeElement> children) {
        Map<Pair<String, Integer>, CallDefinition> functionByNameArity = new HashMap<>(children.size());

        for (TreeElement child : children) {
            if (child instanceof CallDefinition) {
                CallDefinition callDefinition = (CallDefinition) child;

                if (callDefinition.time() == Timed.Time.RUN) {
                    Pair<String, Integer> nameArity = pair(callDefinition.name(), callDefinition.arity());
                    functionByNameArity.put(nameArity, callDefinition);
                }
            }
        }

        return functionByNameArity;
    }

    private static Collection<TreeElement> provideNodesFromChild(@NotNull TreeElement child) {
        Collection<TreeElement> nodes = null;

        if (child instanceof Use) {
            Use use = (Use) child;
            PsiElement[] finalArguments = finalArguments(use.call());

            assert finalArguments != null;

            if (finalArguments.length > 0) {
                PsiElement firstFinalArgument = finalArguments[0];

                if (firstFinalArgument instanceof ElixirAccessExpression) {
                    PsiElement accessExpressionChild = stripAccessExpression(firstFinalArgument);

                    if (accessExpressionChild instanceof QualifiableAlias) {
                        PsiReference reference = accessExpressionChild.getReference();

                        if (reference != null) {
                            PsiElement ancestor = reference.resolve();

                            while (ancestor != null && !(ancestor instanceof PsiFile)) {
                                if (ancestor instanceof Call) {
                                    Call call = (Call) ancestor;

                                    if (Module.is(call)) {
                                        Module module = new Module(call);
                                        Call[] childCalls = macroChildCalls(call);

                                        Map<Pair<String, Integer>, CallDefinition> macroByNameArity = new HashMap<>(childCalls.length);

                                        for (Call childCall : childCalls) {
                                                /* portion of {@link org.elixir_lang.structure_view.element.enclosingModular.Module#childCallTreeElements}
                                                   dealing with macros, restricted to __using__/1 */
                                            if (CallDefinitionClause.isMacro(childCall)) {
                                                Pair<String, IntRange> nameArityRange = CallDefinitionClause.nameArityRange(childCall);

                                                if (nameArityRange != null) {
                                                    String name = nameArityRange.first;
                                                    IntRange arityRange = nameArityRange.second;

                                                    if (name.equals(USING) && arityRange.containsInteger(1)) {
                                                        addClausesToCallDefinition(
                                                                childCall,
                                                                name,
                                                                arityRange,
                                                                macroByNameArity,
                                                                module,
                                                                Timed.Time.COMPILE,
                                                                element -> {
                                                                }
                                                        );
                                                    }
                                                }
                                            }
                                        }

                                        if (macroByNameArity.size() > 0) {
                                            PsiElement[] usingArguments;
                                            CallDefinition macro;
                                            CallDefinitionClause matchingClause = null;

                                            if (finalArguments.length > 1) {
                                                usingArguments = Arrays.copyOfRange(finalArguments, 1, finalArguments.length);
                                                Pair<String, Integer> nameArity = pair(USING, usingArguments.length);
                                                macro = macroByNameArity.get(nameArity);

                                                if (macro != null) {
                                                    matchingClause = macro.matchingClause(usingArguments);
                                                }
                                            } else {
                                                    /* `use <ALIAS>` will calls `__using__/1` even though there is
                                                       no additional argument, but it obviously can't select a clause. */
                                                Pair<String, Integer> nameArity = pair(USING, 1);
                                                macro = macroByNameArity.get(nameArity);
                                                List<CallDefinitionClause> macroClauseList = macro.clauseList();

                                                if (macroClauseList.size() == 1) {
                                                    matchingClause = macroClauseList.get(0);
                                                } else {
                                                    // TODO match default argument clause/head to non-default argument clause that would be executed.
                                                }
                                            }

                                            if (matchingClause != null) {
                                                TreeElement[] callDefinitionClauseChildren = matchingClause.getChildren();
                                                int length = callDefinitionClauseChildren.length;

                                                if (length > 0) {
                                                    TreeElement lastCallDefinitionClauseChild = callDefinitionClauseChildren[length - 1];

                                                    if (lastCallDefinitionClauseChild instanceof Quote) {
                                                        Quote quote = (Quote) lastCallDefinitionClauseChild;
                                                        Quote injectedQuote = quote.used(use);
                                                        TreeElement[] injectedQuoteChildren = injectedQuote.getChildren();
                                                        nodes = new ArrayList<>(injectedQuoteChildren.length);

                                                        for (TreeElement injectedQuoteChild : injectedQuoteChildren) {
                                                            if (!(injectedQuoteChild instanceof Overridable)) {
                                                                nodes.add(injectedQuoteChild);
                                                            }
                                                        }

                                                        break;
                                                    }
                                                }
                                            }
                                        }
                                    } else {
                                        break;
                                    }
                                }

                                ancestor = ancestor.getParent();
                            }
                        }
                    }
                }
            }
        }

        if (nodes == null) {
            nodes = Collections.emptyList();
        }

        return nodes;
    }

    public static Collection<TreeElement> provideNodesFromChildren(@NotNull Collection<TreeElement> children) {
        Collection<TreeElement> nodes = new ArrayList<>();

        for (TreeElement child : children) {
            Collection<TreeElement>  nodesFromChild = provideNodesFromChild(child);
            nodes.addAll(nodesFromChild);
        }

        return nodes;
    }

    /*
     * Instance Methods
     */

    @NotNull
    @Override
    public String getActionIdForShortcut() {
        return "FileStructurePopup";
    }

    @NotNull
    @Override
    public String getCheckBoxText() {
        return "Show Used";
    }

    /**
     * Returns a unique identifier for the action.
     *
     * @return the action identifier.
     */
    @NotNull
    @Override
    public String getName() {
        return ID;
    }

    /**
     * Returns the presentation for the action.
     *
     * @return the action presentation.
     * @see ActionPresentationData#ActionPresentationData(String, String, Icon)
     */
    @NotNull
    @Override
    public ActionPresentation getPresentation() {
        return new ActionPresentationData("Show Used", null, AllIcons.Hierarchy.Supertypes);
    }

    @NotNull
    @Override
    public Shortcut[] getShortcut() {
        throw new IncorrectOperationException("see getActionIdForShortcut()");
    }

    @NotNull
    @Override
    public Collection<TreeElement> provideNodes(@NotNull TreeElement node) {
        Collection<TreeElement> nodes = new ArrayList<>();

        if (node instanceof Module) {
            Module module = (Module) node;
            TreeElement[] children = module.getChildren();
            Collection<TreeElement> childCollection = Arrays.asList(children);

            nodes = provideNodesFromChildren(childCollection);
            nodes = filterOverridden(nodes, childCollection);
        }

        return nodes;
    }

}

File: src/org/elixir_lang/structure_view/node_provider/Used.kt
package org.elixir_lang.structure_view.node_provider

import com.intellij.icons.AllIcons
import com.intellij.ide.util.ActionShortcutProvider
import com.intellij.ide.util.FileStructureNodeProvider
import com.intellij.ide.util.treeView.smartTree.ActionPresentation
import com.intellij.ide.util.treeView.smartTree.ActionPresentationData
import com.intellij.ide.util.treeView.smartTree.TreeElement
import com.intellij.openapi.actionSystem.Shortcut
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFile
import com.intellij.util.IncorrectOperationException
import org.elixir_lang.NameArity
import org.elixir_lang.psi.ElixirAccessExpression
import org.elixir_lang.psi.QualifiableAlias
import org.elixir_lang.psi.call.Call
import org.elixir_lang.psi.impl.call.finalArguments
import org.elixir_lang.psi.impl.call.macroChildCalls
import org.elixir_lang.psi.impl.stripAccessExpression
import org.elixir_lang.semantic.call.definition.clause.Time
import org.elixir_lang.semantic.semantic
import org.elixir_lang.structure_view.element.*
import org.elixir_lang.structure_view.element.call.Definition
import org.elixir_lang.structure_view.element.call.definition.Clause
import org.elixir_lang.structure_view.element.modular.Module
import org.jetbrains.annotations.NonNls
import java.util.*

class Used : FileStructureNodeProvider<TreeElement>, ActionShortcutProvider {
    override fun getActionIdForShortcut(): String = "FileStructurePopup"
    override fun getCheckBoxText(): String = "Show Used"

    /**
     * Returns a unique identifier for the action.
     *
     * @return the action identifier.
     */
    override fun getName(): String = ID

    /**
     * Returns the presentation for the action.
     *
     * @return the action presentation.
     * @see ActionPresentationData.ActionPresentationData
     */
    override fun getPresentation(): ActionPresentation =
        ActionPresentationData("Show Used", null, AllIcons.Hierarchy.Supertypes)

    override fun getShortcut(): Array<Shortcut> = throw IncorrectOperationException("see getActionIdForShortcut()")

    override fun provideNodes(node: TreeElement): Collection<TreeElement> =
        (node as? Module)?.children?.toList()?.let { childCollection ->
            provideNodesFromChildren(childCollection).let { filterOverridden(it, childCollection) }
        } ?: emptyList()

    companion object {
        @NonNls
        const val ID = "SHOW_USED"
        private const val USING = "__using__"

        private fun filterOverridden(
            nodesFromChildren: Collection<TreeElement>,
            children: Collection<TreeElement>
        ): Collection<TreeElement> {
            val childFunctionByNameArity = functionByNameArity(children)

            return nodesFromChildren
                .filterIsInstance<Definition>()
                // only functions work with `defoverridable`
                .filter { it.time == Time.RUN }
                .filterNot { definition ->
                    definition.semantic.nameArityInterval?.let { nameArityInterval ->
                        val name = nameArityInterval.name

                        nameArityInterval.arityInterval.closed().all { arity ->
                            val nameArity = NameArity(name, arity)

                            childFunctionByNameArity.containsKey(nameArity)
                        }
                    } ?: false
                }
        }

        fun functionByNameArity(children: Collection<TreeElement>): Map<NameArity, Definition> =
            children
                .filterIsInstance<Definition>()
                .filter { it.time == Time.RUN }
                .flatMap { definition ->
                    definition
                        .semantic
                        .nameArityInterval?.let { nameArityInterval ->
                            val name = nameArityInterval.name

                            nameArityInterval.arityInterval.closed().map { arity ->
                                NameArity(name, arity) to definition
                            }
                        }
                        .orEmpty()
                }
                .toMap()

        private fun provideNodesFromChild(child: TreeElement): Collection<TreeElement> {
            var nodes: MutableCollection<TreeElement>? = null

            if (child is Use) {
                val finalArguments = child.call().finalArguments()!!

                if (finalArguments.isNotEmpty()) {
                    val firstFinalArgument = finalArguments[0]

                    if (firstFinalArgument is ElixirAccessExpression) {
                        val accessExpressionChild = firstFinalArgument.stripAccessExpression()

                        if (accessExpressionChild is QualifiableAlias) {
                            val reference = accessExpressionChild.getReference()

                            if (reference != null) {
                                var ancestor = reference.resolve()

                                while (ancestor != null && ancestor !is PsiFile) {
                                    if (ancestor is Call) {
                                        val call = ancestor
                                        val semantic = call.semantic

                                        if (semantic is org.elixir_lang.semantic.Module) {
                                            val module =
                                                Module(semantic = semantic as org.elixir_lang.semantic.Modular)
                                            val childCalls = call.macroChildCalls()

                                            val macroByNameArity = HashMap<NameArity, Definition>(childCalls.size)

                                            for (childCall in childCalls) {
                                                val childSemantic = childCall.semantic

                                                /* portion of {@link org.elixir_lang.structure_view.element.enclosingModular.Module#childCallTreeElements}
                                                   dealing with macros, restricted to __using__/1 */
                                                if (childSemantic
                                                            is org.elixir_lang.semantic.call.definition.Clause
                                                ) {
                                                    val nameArityInterval = childSemantic.nameArityInterval

                                                    if (nameArityInterval != null) {
                                                        val name = nameArityInterval.name
                                                        val arityInterval = nameArityInterval.arityInterval

                                                        if (name == USING && arityInterval.contains(1)) {
                                                            TODO()
                                                        }
                                                    }
                                                }
                                            }

                                            if (macroByNameArity.size > 0) {
                                                val usingArguments: Array<PsiElement>
                                                val macro: Definition?
                                                var matchingClause: Clause? = null

                                                if (finalArguments.size > 1) {
                                                    usingArguments =
                                                        Arrays.copyOfRange(finalArguments, 1, finalArguments.size)
                                                    val nameArity = NameArity(USING, usingArguments.size)
                                                    macro = macroByNameArity[nameArity]

                                                    if (macro != null) {
                                                        matchingClause = macro.matchingClause(usingArguments)
                                                    }
                                                } else {
                                                    /* `use <ALIAS>` will calls `__using__/1` even though there is
                                                       no additional argument, but it obviously can't select a clause. */
                                                    val nameArity = NameArity(USING, 1)
                                                    macro = macroByNameArity[nameArity]
                                                    val macroClauseList = macro!!.clauseList()

                                                    if (macroClauseList.size == 1) {
                                                        matchingClause = macroClauseList[0]
                                                    } else {
                                                        // TODO match default argument clause/head to non-default argument clause that would be executed.
                                                    }
                                                }

                                                if (matchingClause != null) {
                                                    val callDefinitionClauseChildren = matchingClause.children
                                                    val length = callDefinitionClauseChildren.size

                                                    if (length > 0) {
                                                        val lastCallDefinitionClauseChild =
                                                            callDefinitionClauseChildren[length - 1]

                                                        if (lastCallDefinitionClauseChild is Quote) {
                                                            val injectedQuote =
                                                                lastCallDefinitionClauseChild.used(child)
                                                            val injectedQuoteChildren = injectedQuote.children
                                                            nodes = ArrayList(injectedQuoteChildren.size)

                                                            for (injectedQuoteChild in injectedQuoteChildren) {
                                                                if (injectedQuoteChild !is Overridable) {
                                                                    nodes.add(injectedQuoteChild)
                                                                }
                                                            }

                                                            break
                                                        }
                                                    }
                                                }
                                            }
                                        } else {
                                            break
                                        }
                                    }

                                    ancestor = ancestor.parent
                                }
                            }
                        }
                    }
                }
            }

            if (nodes == null) {
                nodes = mutableListOf()
            }

            return nodes
        }

        fun provideNodesFromChildren(children: Collection<TreeElement>): Collection<TreeElement> =
            children.flatMap { provideNodesFromChild(it) }
    }
}
